src/Symfony/Component/Security/Core/Tests/Authorization/ExpressionLanguageTest.php

```diff
@@ -52,7 +51,6 @@ public function provider()
         $user = new InMemoryUser('username', 'password', $roles);
 
         $noToken = null;
-        $anonymousToken = new AnonymousToken('firewall', 'anon.');
         $rememberMeToken = new RememberMeToken($user, 'firewall-name', 'firewall');
         $usernamePasswordToken = new UsernamePasswordToken('username', 'password', 'firewall-name', $roles);
```

Comment:
We should probably deprecate unauthenticated tokens and non-UserInterface token users. The use-cases for these have been replaced with Passports.

Reply:
Could you please elaborate? What's the before/after?

Reply:
Before, e.g. a form login would create `UsernamePasswordToken('the-input-username', 'the-input-password')` and a new token would be created by the DaoAuthenticationProvider `UsernamePasswordToken($userFromUserProvider, null, $userFromUserProvider->getRoles())`.

Tokens use all sorts of different conditions to mark themselves as "authenticated" (after provider) or "unauthenticated" (before provider).

However, all unauthenticated token use-cases have been replaced with passports. This means that if you have a token, it should always be authenticated. We don't yet validate for this in 5.4+, so it might be unauthenticated.

My proposal:
* Deprecate unauthenticated tokens + `TokenInterface::isAuthenticated()`
* Deprecate not passing a `UserInterface` instance to tokens (simplifying `TokenInterface::getUser()` from `string|Stringify|UserInterface` to just `UserInterface`)

Reply:
👍 Your proposal makes sense to me. I had a quick look at the implications for both changes and it seems doable with not much hassle.

Reply:
#42050